### PR TITLE
Optimize pd monitor configuration write logic.

### LINKED ISSUE #3292: Refactor Tidbmonitor design.
## Description

Currently, there are some problems with TiDBMonitor. We organize the requirements and list the requirements that need improvement.

1. TiDBMonitor is a stateful application but is created as a deployment. We should switch to StatefulSet. #2968
2. TiDBMonitor monitors multiple TiDB clusters for both TLS and non-TLS environment.
    After discussion, we will not support monitoring multiple clusters with one TidbMonitor when TLS enabled because this would require the following restrictions:
* All of the monitored clusters have to enable TLS
* Users must ensure that all the monitored clusters use the same CA certificate
* A dedicated cluster client secret needs to be created for TidbMonitor

The above restrictions are not reasonable so we will just support monitoring multiple clusters with TLS disabled with one TidbMonitor.
3. Monitor multiple TiDB Clusters with users own Prometheus #3184
    We can provide an example with Prometheus Operator.
4. Aggregate the monitoring data of multiple TiDB Clusters, we can refer to the Prometheus operator and support the Thanos spec.  #3118 #3117
5. Combination of Kubernetes monitoring and TiDB cluster monitoring. #3099
6. add external fields for `kubectl get tm` command. #2169 #2982
7. Grafana dashboards for multiple clusters.
8. The PD Dashboard monitor address is currently TidbCluster controller updated, it should be updated by tidbmonitor controller. And PD dashboard doesn't support multiple clusters.
9. Prometheus need to reload configuration  when spec filed cause configMap update .

   -  We can regenerate configMap name cause deployment rolling update.
   - Hot update configuration through reloader and call prometheus reload api (More user-friendly).

## Category
Feature

## Value
Refactor tidbMonitor.

## TODO lists
- P0 features
	- [x] Tidbmonitor monitors multiple clusters across multiple ns in tls and non-tls envrionments. #3307 
	- [x] Support generating the Grafana dashboard of multiple clusters.
- P1 features	
	- [x] Smooth upgrade deployment to StatefulSet. #2968
	- [x] Support Thanos spec and optimize Thanos example. #3553
	- [ ] Optimize TiDB service, more friendly support with Prometheus operator `ServiceMonitor`.
	- [x] Refactor for PD dashboard address writing.  #3924
	- [x] TidbMonitor multiple replica. #2968 #3815 sloved.

- P2 feature
	- [ ] Fix the combination of Kubernetes monitoring and TiDB cluster monitoring.
	- [x] Add more external fields for `kubectl get tm` command. #3958

Extra features:

- [x] support monitor multiple dm across multiple ns in tls and non-tls environments.
- [x] No need to replace the cluster name in https://github.com/pingcap/monitoring/blob/master/platform-monitoring/operator/init.sh
- [x] The tc parameters in the various functions in pkg/monitor/monitor/monitor_manager.go and pkg/monitor/monitor/util.go can be removed.
- [ ] Optimize `createorupdateSecret` method,reduce interaction with apiserver.
## Workload Estimation (P0 features)
30

## Time
GanttStart: 2020-10-01
GanttDue: 2020-12-31

## Project

### PR DESCRIPTION
**Pull request by mikechengwei**
### What problem does this PR solve?
Fix #3292 .
Optimize pd monitor configuration write logic .TidbMonitor write monitor configuration to  pd etcd directly.

### Code changes

- [ ] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [ ] Manual test 
- [ ] No code 

### Release Notes

```release-note
TidbMonitor write monitor configuration to  pd etcd directly.
```

### CODE REVIEW

**Review by july2993 — APPROVED** [review]

File: pkg/manager/member/tidbcluster_status_manager.go [hunk 1]
```diff
@@ -45,86 +37,9 @@ func (m *TidbClusterStatusManager) Sync(tc *v1alpha1.TidbCluster) error {
 }
 
 func (m *TidbClusterStatusManager) syncTidbMonitorRefAndKey(tc *v1alpha1.TidbCluster) error {
```
Comment (DanielZhangQD) [inline pkg/manager/member/tidbcluster_status_manager.go:39]: Just remove this function and move L40 to L36.

File: pkg/monitor/monitor/monitor_manager.go [hunk 2]
```diff
@@ -19,6 +19,8 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/pingcap/tidb-operator/pkg/pdapi"
+
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:23]: Remove this.

File: pkg/monitor/monitor/monitor_manager.go [hunk 3]
```diff
@@ -84,23 +89,10 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
...
-		// been set by another TidbMonitor.
-		// Patch tidbcluster status first to avoid multi tidbmonitor monitoring the same tidbcluster
-		if !tc.IsHeterogeneous() {
-			if err := m.patchTidbClusterStatus(tc, monitor); err != nil {
-				message := fmt.Sprintf("Sync TidbMonitorRef into targetCluster[%s/%s] status failed, err:%v", tc.Namespace, tc.Name, err)
-				klog.Error(message)
-				m.deps.Recorder.Event(monitor, corev1.EventTypeWarning, FailedSync, err.Error())
-				return err
-			}
+		err = m.syncDashboardMetricStorage(tc, monitor)
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:92]: We only need to sync dashboard config for non Heterogeneous tc.

File: pkg/monitor/monitor/monitor_manager.go [hunk 4]
```diff
@@ -84,23 +89,10 @@ func (m *MonitorManager) SyncMonitor(monitor *v1alpha1.TidbMonitor) error {
...
+		if err != nil {
+			klog.Errorf("tidb[%s/%s] update dashboard pd monitor metrics error: %v", tc.Namespace, tc.Name, err)
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:94]:
```suggestion
			klog.Errorf("Fail to sync TiDB Dashboard metrics config for TiDB cluster [%s/%s], error: %v", tc.Namespace, tc.Name, err)
```

File: pkg/monitor/monitor/monitor_manager.go [hunk 5]
```diff
@@ -613,3 +581,79 @@ func (m *MonitorManager) patchPVClaimRef(pv *corev1.PersistentVolume, patchPvcNa
...
+	if err != nil {
+		return err
+	}
+	if tc.IsTLSClusterEnabled() {
+		defer pdEtcdClient.Close()
+	}
+
+	grafanaExist := tm.Spec.Grafana != nil
+	// sync prometheus key
+	err = syncComponent(true, tm, prometheusComponent, 9090, pdEtcdClient)
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:600]: How can we clear the keys in the etcd if the TidbMonitor CR is deleted?

File: pkg/monitor/monitor/monitor_manager.go [hunk 6]
```diff
@@ -700,3 +668,80 @@ func (m *MonitorManager) syncAssetSecret(monitor *v1alpha1.TidbMonitor, store *S
...
+func (m *MonitorManager) syncDashboardMetricStorage(tc *v1alpha1.TidbCluster, tm *v1alpha1.TidbMonitor) error {
+	if tc.Spec.PD == nil {
+		return nil
+	}
+	pdEtcdClient, err := m.deps.PDControl.GetPDEtcdClient(pdapi.Namespace(tc.Namespace), tc.Name, tc.IsTLSClusterEnabled())
+
+	if err != nil {
+		return err
+	}
+	if tc.IsTLSClusterEnabled() {
```
Comment (csuzhangxc) [inline pkg/monitor/monitor/monitor_manager.go:681]: no need to check `tc.IsTLSClusterEnabled()` after #3929. see https://github.com/pingcap/tidb-operator/pull/3929/files#diff-a8fc0b3952619cae74a70fd53685f27e760ccb52a03682eff264c51f5fefdc47R116

File: pkg/monitor/monitor/monitor_manager.go [hunk 7]
```diff
@@ -700,3 +668,80 @@ func (m *MonitorManager) syncAssetSecret(monitor *v1alpha1.TidbMonitor, store *S
...
+			return err
+		}
+		//10 min
+		err = etcdClient.PutTTLKey(key, v, 600)
+		if err != nil {
+			klog.Error(err.Error())
+			return err
+		}
+	} else {
+		err := etcdClient.DeleteKey(key)
```
Comment (DanielZhangQD) [inline pkg/monitor/monitor/monitor_manager.go:715]: Is this necessary since we have TTL set in L709?

File: pkg/pdapi/pdetcd.go [hunk 8]
```diff
@@ -93,6 +95,14 @@ func (c *pdEtcdClient) PutKey(key, value string) error {
 	return err
 }
 
+func (c *pdEtcdClient) PutTTLKey(key, value string, ttl int64) error {
+	ctx, cancel := context.WithTimeout(context.Background(), c.timeout)
+	defer cancel()
+	lease, _ := c.etcdClient.Grant(ctx, ttl)
```
Comment (DanielZhangQD) [inline pkg/pdapi/pdetcd.go:101]: Why ignore the error here?

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**